File: src/main/java/com/guhungry/prime/PrimeFactor.java
package com.guhungry.prime;

import java.util.ArrayList;
import java.util.List;

public class PrimeFactor {
    public static List<Integer> of(int n) {
        List<Integer> factors = new ArrayList<Integer>();
        int max = (int)Math.sqrt(n);
        for (int divisor = 2; n > 1 && divisor <= max; divisor += divisor > 2 ? 2 : 1)
            for (; n % divisor == 0; n /= divisor)
                factors.add(divisor);
        if (n > 1) factors.add(n);
        return factors;
    }
}

File: src/main/java/com/guhungry/prime/PrimeFactor.kt
package com.guhungry.prime

fun Int.primeFactors(): IntArray {
    val factors = arrayListOf<Int>()
    var n = this
    val max = kotlin.math.sqrt(n.toDouble()).toInt()
    for (divisor in 2 .. max)
        while (n % divisor == 0) {
            factors.add(divisor)
            n /= divisor
        }
    if (n > 1) factors.add(n)
    return factors.toIntArray()
}